tools/src/postgen-bytetypes.ts

```diff
@@ -0,0 +1,86 @@
+const pascalcase = require('pascalcase');
+
+/**
+ * This file converts string types with:
+ * - byte data
+ * to
+ * - byte types
+ *
+ * ## Example
+ * ### Before
+ * type Message struct {
+ *   Data        *string        `json:"data,omitempty"`        // The binary data in the message.
+ * }
+ * 
+ * ### After
+ * type Message struct {
+ *   Data        []byte        `json:"data,omitempty"`        // The binary data in the message.
+ * }
+ * 
+ * Also handles repeated proto fields.
+ * 
+ * ---
+ * 
+ * Does this by inspecting the proto, and modifing the source where appropriate.
+ * 
+ * This modification currently cannot be done at the JSON schema + Quicktype level,
+ * as JSON schema + Quicktype does create base64 formatted strings.
+ * @see https://github.com/quicktype/quicktype/issues/138
+ * @param {string} golangFile The golang file source code
+ * @param {string} protoFile The proto source
+ * @returns {string} The golang file with fixed base64 fields
+ */
+export const fixByteFields = (golangFile: string, protoFile: string) => {
+  // A proto line with bytes
+  type LineWithByte = {
+    field: string;
+    isRepeated: boolean;
+  };
+  
+  // Get all proto lines with the "bytes" type:
```

Comment:
A more reliable alternative to this would be to use the compiled proto descriptor. I don't know what the Go protobuf library is like, but certainly the C# equivalent wouldn't be terribly complicated. (We use the compiled proto descriptor to extract CloudEvent types in C#, for example.)

I can imagine all kinds of places where this could go wrong, e.g. by parsing a comment as if it were a field declaration.

I'd be okay with this as a temporary measure if we believe that the schema generator + QuickType are likely to handle this properly in the future, but I wouldn't want this to be in place for the long term.

Reply:
> A more reliable alternative to this would be to use the compiled proto descriptor. I don't know what the Go protobuf library is like, but certainly the C# equivalent wouldn't be terribly complicated. (We use the compiled proto descriptor to extract CloudEvent types in C#, for example.)

I've thought about alternative solutions to achieve this Go behavior and chose this route as the best design all things considered. A long-term solution would be to change a couple things: the JSON schema to use the new (draft 7) `'contentEncoding': 'base64'`and the Golang generator to read this field and apply the correct type. However, that isn't in place today. There are a couple downstream libraries that need updates in order to support this.

We could use the `@grpc/proto-loader` for the proto part of this fix to use a compiled proto descriptor, but that wouldn't be able to modify the generated Go source code and would have a minor benefit. 

> I can imagine all kinds of places where this could go wrong, e.g. by parsing a comment as if it were a field declaration.

This isn't AST parsing; yes there are places where it could go wrong that we prevent. We could log out the ~7 places this fix modifies the Go byte fields. I have tried to make the post-generation step as readable as possible.

> I'd be okay with this as a temporary measure if we believe that the schema generator + QuickType are likely to handle this properly in the future, but I wouldn't want this to be in place for the long term.

Yes, the goal would be for better schemas and generators for base64 data. This requires a few upstream changes like https://github.com/quicktype/quicktype/issues/138.